pkg/services/featuremgmt/manager.go

```diff
@@ -0,0 +1,236 @@
+package featuremgmt
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/grafana/grafana/pkg/infra/log"
+
+	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/models"
+)
+
+type FeatureManager struct {
+	isDevMod  bool
+	licensing models.Licensing
+	flags     map[string]*FeatureFlag
+	enabled   map[string]bool // only the "on" values
+	toggles   *FeatureToggles
+	config    string // path to config file
+	vars      map[string]interface{}
+	log       log.Logger
+}
+
+// This will merge the flags with the current configuration
+func (fm *FeatureManager) registerFlags(flags ...FeatureFlag) {
+	for idx, add := range flags {
+		if add.Name == "" {
+			continue // skip it with warning?
+		}
+		flag, ok := fm.flags[add.Name]
+		if !ok {
+			fm.flags[add.Name] = &flags[idx]
+			continue
+		}
+
+		// Selectivly update properties
+		if add.Description != flag.Description {
+			flag.Description = add.Description
+		}
+		if add.DocsURL != flag.DocsURL {
+			flag.DocsURL = add.DocsURL
+		}
+		if add.Expression != flag.Expression {
+			flag.Expression = add.Expression
+		}
+
+		// The least stable state
+		if add.State > flag.State {
+			flag.State = add.State
+		}
+
+		// Only gets more restrictive
+		if add.RequiresDevMode {
+			flag.RequiresDevMode = true
+		}
+
+		if add.RequiresLicense {
+			flag.RequiresLicense = true
+		}
+
+		if add.RequiresRestart {
+			flag.RequiresRestart = true
+		}
+	}
+}
+
+func (fm *FeatureManager) evaluate(ff *FeatureFlag) bool {
+	if ff.RequiresDevMode && !fm.isDevMod {
+		return false
+	}
+
+	if ff.RequiresLicense && !fm.licensing.FeatureEnabled(ff.Name) {
```

Comment:
@DanCech -- this will make sure any features that require a license are checked with the license provider